### ISSUE TITLE
Access Response object in the Scrapy Pipeline

### ISSUE THREAD

**Issue by jacob1237** [scrapy/scrapy#2436]
Hello,

I'd like to implement a pipeline which compares a checksum of the current item with some old values in cache and drops the items that are not modified.
Looks like this is a standard use-case for change detection systems (e.g. price monitoring).

Right now I put the cache (old checksums) into the Request.meta field, but there is no way to access it inside the pipeline.

As a workaround I did the same trick with the Spider Middleware (process_spider_output), but now I have another problem: when I want to skip processing of some item and signalize other components about that with Exception (like DropItem), to catch it with the **spider_error** signal, I can't suppress the default exception handler.

So every time I always get some tracebacks in a log (I don't want them because I have my own spider exception handler).

This is because of that code in `scrapy/core/scraper.py`:
```python
def handle_spider_error(self, _failure, request, response, spider):
    exc = _failure.value
    if isinstance(exc, CloseSpider):
        self.crawler.engine.close_spider(spider, exc.reason or 'cancelled')
        return
    referer = request.headers.get('Referer')
    logger.error(
        "Spider error processing %(request)s (referer: %(referer)s)",
        {'request': request, 'referer': referer},
        exc_info=failure_to_exc_info(_failure),
        extra={'spider': spider}
    )
    self.signals.send_catch_log(
        signal=signals.spider_error,
        failure=_failure, response=response,
        spider=spider
    )
    ...other code here...
```
It seems there is no way to suppress Spider Exceptions before the signal execution (even if I have a handler).

I know that I can implement a custom crawler signal and execute it from my middleware (as a workaround), but won't it be easier to access the Response object from the Pipeline handlers?

I can try to make a patch which will add the additional parameter to the pipeline handler (without breaking compatibility with an old code), but first I'd like to discuss it with the community.

Is it conceptually wrong to access the response object from the pipeline?

**Comment by IAlwaysBeCoding** [comment]
If you do have the `response` being passed to any `Item Pipeline` , then you will have a very long memory leak. Remember that every `Response` has an associated `Request` attach on the `request` attribute. Those responses will have to be maintained and not discarded in order for them to be alive when you pass them through the pipeline. At the moment as soon as they are done passing through the Spider Middlewares then it takes a while after that for the garbage collector to come and pick them up. 

That is a really bad design decision to hold on to the `Response` objects just so you can pass it on a `Item Pipeline`.

**Comment by jacob1237** [comment]
Well, looks like it's not true. At least there is no direct way to produce such "memory leak".
Here is the code from `scrapy/core/scraper.py`:
```python
def _process_spidermw_output(self, output, request, response, spider):
    """Process each Request/Item (given in the output parameter) returned
    from the given spider
    """
    if isinstance(output, Request):
        self.crawler.engine.crawl(request=output, spider=spider)
    elif isinstance(output, (BaseItem, dict)):
        self.slot.itemproc_size += 1
        dfd = self.itemproc.process_item(output, spider)
        dfd.addBoth(self._itemproc_finished, output, response, spider)
        return dfd
    <other code...>
```
You can see that `self.itemproc.process_item` (Deferred) is invoked without `response`, but there is also a callback assigned to that deferred, called `self._itemproc_finished`, which is invoked only AFTER pipeline processing (by Twisted). And yes, it has a `response` object as a parameter.

Also, this method (`self._itemproc_finished`) produces `item_scraped` signal. It means that the `request` lifecycle is longer than you think in your message.
So, for example, I can produce the same "memory leak" from any `Extension` that is listening for `item_scraped` signal.

The only disadvantage I see is that if I'll add the `response` object to the `self.itemproc.process_item`, it will be inconsistent with similar function signatures: the `response` object will be passed in the end of the arguments list in order to maintain compatibility with an old code.

**Comment by pawelmhm** [comment]
@jacob1237 

> I'd like to implement a pipeline which compares a checksum of the current item with some old values in cache and drops the items that are not modified.

Sorry just wanted to clarify one thing. Why do you need response in pipeline if you say you want to compare items?

> Right now I put the cache (old checksums) into the Request.meta field, but there is no way to access it inside the pipeline.

if checksum is tied to Scrapy.Item why are you putting it in request meta instead of putting it in item as a field? Just add field 'checksum' to your item, keep some old item somewhere (e.g. in some database that can be accessed from pipeline) and compare checksums in `process_item` pipeline method.

There is Scrapy extension called Delta-Fetch that seems to be something that might be useful for you, maybe you can give it a try https://github.com/scrapy-plugins/scrapy-deltafetch

If you actually really need response in pipeline (very rare but happens), you can just do following: store response on disk or in database in spider middleware `process_spider_output` method - generate 'filename' key add it to item. In pipeline when processing item, fetch 'filename' key from item and use it to load response from filesystem.

**Comment by jacob1237** [comment]
@pawelmhm 

> Sorry just wanted to clarify one thing. Why do you need response in pipeline if you say you want to compare items?

I need only the `meta` field where I store old checksums.

Also, hypothetically (not my case, just example) I may want to store some response-specific parameters such as URL, status or some meta fields, together with the Item (because pipelines can be used for item storing).

> if checksum is tied to Scrapy.Item why are you putting it in request meta instead of putting it in item as a field? Just add field 'checksum' to your item, keep some old item somewhere (e.g. in some database that can be accessed from pipeline) and compare checksums in process_item pipeline method.

Because I have no elegant way to access old checksum (from pipeline) except the database retrieval request. So if the page contains 100 items (in our case it can be up to 500), I will need to make 100 retrieval requests to the database. It is OK for simple cases when you can use local Berkeley DB files (like **DeltaFetch** do), but it is not good for distributed systems with centralized network-access storages (our case).

So in our case it is better to retrieve checksums once and put them into the `meta` field than produce more than 100 additional retrieval requests to the database for every URL.

> There is Scrapy extension called Delta-Fetch that seems to be something that might be useful for you, maybe you can give it a try https://github.com/scrapy-plugins/scrapy-deltafetch

Thanks for advice! Actually, I've already solved my problem and DeltaFetch wasn't the better solution for our case (because it doesn't detect any real changes in Item fields).

> If you actually really need response in pipeline (very rare but happens), you can just do following: store response on disk or in database in spider middleware process_spider_output method - generate 'filename' key add it to item. In pipeline when processing item, fetch 'filename' key from item and use it to load response from filesystem.

Of course it's possible, but it looks like a dirty hack for such case. That's why I started this discussion.

**Comment by pawelmhm** [comment]
> I need only the meta field where I store old checksums.

ok, so you get checksum of scraped item in request.meta. So, why not add old checksum to new item and compare it in pipeline with new checksum in pipeline?

> Also, hypothetically (not my case, just example) I may want to store some response-specific parameters such as URL, status or some meta fields, together with the Item (because pipelines can be used for item storing).

hmm why not store them in item too? Keeping whole response body in memory (which may be huge) just to get status code or other parameter seems excessive. You can use spider middleware [process_spider_output](https://doc.scrapy.org/en/latest/topics/spider-middleware.html#scrapy.spidermiddlewares.SpiderMiddleware.process_spider_output), it has access to response and all output from spider callback and is ideal for this kind of use case, you can fetch something from response and append to item as a field.

> Because I have no elegant way to access old checksum (from pipeline) except the database retrieval request.

well, how are you accessing old checksum that you get in request.meta? Will accessing it in pipeline differ from accessing it in spider? 

> I will need to make 100 retrieval requests to the database

yeah you're right, making database query for each item is not efficient. You can bypass this easily by downloading all checksums in pipeline on spider_opened signal. 

>  Actually, I've already solved my problem

that's good. 

I guess your question is about adding new functionality to Scrapy - accessing response from item pipeline. It sounds like good question, but it seems like rather radical and little risky change. I'm not sure if there are good arguments for it. I'm not really convinced by your arguments. The fact that you managed to solve your problem without this change seems to confirm that it's not very urgent to add it, am I right?

**Comment by jacob1237** [comment]
@pawelmhm

> hmm why not store them in item too? Keeping whole response body in memory (which may be huge) just to get status code or other parameter seems excessive.

This is not a problem because response body is not destroying during the pipeline processing. The last reference to the response (and its body) MAY be destroyed only after `item_scraped` signal call (check [this answer](https://github.com/scrapy/scrapy/issues/2436#issuecomment-266983206) about the response lifecycle).

So the response body is keeping anyway, even if you don't use it in your pipeline.

> You can use spider middleware process_spider_output, it has access to response and all output from spider callback and is ideal for this kind of use case, you can fetch something from response and append to item as a field.

Well, this is my current solution. The problem occurs when I want to drop some items (which weren't modified). The `DropItem` exception can be raised only from the pipeline.

So when I want to drop an item from the middleware, I can't do that in a direct way. Right now I manually call the `item_dropped` signal to be able to catch it from extensions (`crawler.signals.send_catch_log_deferred()`).

This looks like a duct tape programming.

> yeah you're right, making database query for each item is not efficient. You can bypass this easily by downloading all checksums in pipeline on spider_opened signal.

Not my case. We have long-running Scrapy workers that are feeding from the one distributed queue in a real-time. This is also the [Frontera](https://github.com/scrapinghub/frontera) case (distributed mode).

So as I said before, we don't have any ability to do that from the pipeline.

> The fact that you managed to solve your problem without this change seems to confirm that it's not very urgent to add it, am I right?

It is not very urgent, you're right. I've started this discussion both to show and to better understand the problem.

And if we'll come to any decision, I can make a pull request to fix that problem.

**Comment by NewUserHa** [comment]
pass some meta/additional msg/variable to item pipeline is not a bad idea. I can image there're cases where someone want to do some `if....:` without mess up with the well prepared items.

**Comment by mredaelli** [comment]
Just found this after posting [a question on SO](https://stackoverflow.com/questions/59748229/access-request-in-an-item-pipeline).

Let me add my use case in favour of adding a `request` parameter for pipelines: validation.

As I wrote there, we have a validation item pipeline, and we want now to extend it so that one field of the item (there is only one type of items) is sometimes not required anymore.

It could be done with a custom setting of the spider, but then it would apply to all the items of the spider; however, we want to control this on a per-item basis.

If this were a spider middleware, we'd add a meta key to the request, and all would be well.

As it is now, however, all I can think of would be either:

   -  to add a meta field to the item itself, which I would very much like to avoid: the item is the item, not its metadata -- that's explicitly the role of `meta` :)
   - to transform this to a spidermiddleware, which, besides not being the "usual" stage for validation as per documentation, would require to move other pipelines too.

**Comment by mredaelli** [comment]
On the other hand, it sounds right that conceptually the request is not needed anymore in the item pipeline (since after all you cannot do more requests).

So how about passing not the request, but only `request.meta` to the pipeline, along with the item?

**Comment by elacuesta** [comment]
I believe it's technically possible to pass the request to the `process_item` methods, since we do have access to it when the methods are invoked ([source](https://github.com/scrapy/scrapy/blob/1.8.0/scrapy/core/scraper.py#L179-L197)).
Whether or not it's a good idea to introduce such change, which would be backward-incompatible, is a different thing. Personally, I'm inclined to say "no". I think this is not what the item pipeline is meant to be and, as mentioned already, the spider middleware feature already provides access to both produced items and the request from which they originated.

**Comment by mredaelli** [comment]
I tend to agree that adding `request` is out of scope. However, the documentation for `Request.meta` says:

> This dict is empty for new Requests, and is usually populated by different Scrapy components (extensions, middlewares, etc). So the data contained in this dict depends on the extensions you have enabled.

There's an "etc", and pipelines are not explicitly excluded ;) I personally don't see why they should be excluded, and can see at least two good use cases in this thread.

On the other hand, I understand the backward incompatibility issue (but we have a good trace to follow in #3682, thanks for pointing it out!)

**Comment by AdrianAtZyte** [comment]
It sounds to me like spider middlewares are the way to go, and their only shortcomings are things like not triggering the same signals and stats as an item pipeline. I wonder if we could address the shortcomings by allowing to yield DropItem from a spider middleware. But I also wonder if that’s even worth the effort, or we should just close this.
